### PR TITLE
Scripts: Update stylelint dependency and the default config

### LINKED ISSUE #48142: @wordpress/stylelint-config uses outdated stylelint (14.2)
### Description

The @wordpress/stylelint-config NPM modules has an outdated stylelint dependency (14.2). Current version is 15.1. This causes a dependency conflict for folks using the module.

### Step-by-step reproduction instructions

n/a

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### LINKED ISSUE #54127: [@wordpress/stylelint-config] The premise of "@21.24.0", "Stylelint" is out of date.
### Description

The premise of "@wordpress/stylelint-config@21.24.0", "Stylelint", is out of date. Wouldn't it be better to at least cover "stylelint@"^15.10.3"?

### Step-by-step reproduction instructions

none

### Screenshots, screen recording, code snippet

image

## System:
 - OS: macOS 13.5.1
 - CPU: (12) x64 Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
 - Memory: 219.47 MB / 32.00 GB
 - Shell: 3.2.57 - /bin/bash
## Binaries:
 - Node: 20.5.1 - ~/.nodebrew/current/bin/node
 - Yarn: 1.22.19 - ~/.nodebrew/current/bin/yarn
 - npm: 10.0.0 - ~/.nodebrew/current/bin/npm
 - pnpm: 8.6.12 - ~/.nodebrew/current/bin/pnpm
## npmPackages:
 - @wordpress/scripts: ^26.12.0 => 26.12.0

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by mikeybinns**
## What?
Updating the required version of Stylelint to the latest version, and ensuring we aren't removing any rules which were migrated to a new package.

## Why?
Keeping the Stylelint dependency up to date is important so we continue to get support and related tooling continues to work.
It's also important that the WordPress config remains the same so there aren't issues with CSS across the large codebase.

## How?

- I've updated the stylelint peer dependency from ^14.2 to ^16.8.2 in the `@wordpress/stylelint-config` package
- I've updated the stylelint-config-recommended dependency from ^6.0.0 to ^14.0.1 in the `@wordpress/stylelint-config` package
- I've updated the stylelint-config-recommended-scss dependency from ^5.0.2 to ^14.1.0 in the `@wordpress/stylelint-config` package
- I've added the `@stylistic/stylelint-config` dependency at version ^2.0.0 in the `@wordpress/stylelint-config` package
- I've migrated all stylistic rules to the `@stylistic` namespace in configs and disable/enable comments
- I've moved all stylistic rules into new configs so it is easy for consumers to opt out of stylistic rules instead of using something like `stylelint-config-prettier` to disable them.
- I fixed a bug with the `declaration-block-no-duplicate-properties` rule where it was being overwritten by the `stylelint-config-recommended-scss` config in the scss config.
- I've disabled all newly added rules, even if they would have been added in the recommended configs to make this migration easier (we can add new rules back later if deemed necessary in a MINOR release)
- I've migrated `scss/at-import-partial-extension` to `scss/load-partial-extension` as the former was deprecated.

Once the `@wordpress/stylelint-config` package was updated, I had to make some changes to the `@wordpress/scripts` package and the root stylelint config as the scripts package directly uses the stylelint config, and scripts is used to lint all CSS in the monorepo using the root config.
- I updated the Stylelint dependency from ^14.2 to ^16.8.2
- I updated the scripts default config to use `@wordpress/stylelint-config/scss-stylistic` to keep using the stylistic rules
- I updated the monorepo stylelint config to use `@wordpress/stylelint-config/scss-stylistic` and migrate and rule overrides to the new rule names.

## Testing Instructions
1. Run `npm run lint:css` to run the linting against the monorepo
2. Run `npm run test:unit -- --testPathPattern packages/stylelint-config` to run the automated tests in only this package

Fixes: #48142
Fixes: #54127
Closes: #48143 
Closes: #50336 
Closes: #48769
Closes: #51192
Closes: #54935

### CODE REVIEW

**Comment by mikeybinns** [comment]
Converting to draft while I fix Stylelint unit tests :)

**Comment by mikeybinns** [comment]
@gziolo I'm struggling with fixing these tests. 

It seems that as of v16, Stylelint in Jest must be run in the `node` environment to work in Jest (source: https://github.com/stylelint/stylelint/issues/7894) as the presence of a document breaks things.

I tried running the following command to test if that fix worked:
```sh
npm run test:unit -- --testPathPattern packages/stylelint-config --env=node
```
but this encounters a new error around global setup in `jest-preset-default`.
image

If I delete the contents of the `packages/jest-preset-default/scripts/setup-globals.js` and `test/unit/config/global-mocks.js` files, the tests proceed as expected (snapshot failures).

My jest knowledge isn't very good, and I certainly don't want to go messing around with the jest config, so would you be able to help me out with this? Perhaps there's a way to define a jest config just for this package?

I've pushed all my changes so you can pull down my branch.

**Comment by gziolo** [comment]
https://jestjs.io/docs/configuration#testenvironment-string

Have you tried changing the test env for these test files in the package? That would be the correct environment anyway there. A code comment at the top of every failing test suite file should do the trick. 

```js
/**
 * @jest-environment node
 */
```

All the mocks that depend on `window` can be guarded with the if check to account for the `node` environment where `window` is `undefined` anyway. For example:

```diff
diff --git a/packages/jest-preset-default/scripts/setup-globals.js b/packages/jest-preset-default/scripts/setup-globals.js
index 37922a3302..242d525936 100644
--- a/packages/jest-preset-default/scripts/setup-globals.js
+++ b/packages/jest-preset-default/scripts/setup-globals.js
@@ -2,56 +2,63 @@
 // eslint-disable-next-line @wordpress/wp-global-usage
 globalThis.SCRIPT_DEBUG = true;
 
-// These are necessary to load TinyMCE successfully.
-global.URL = window.URL;
-global.window.tinyMCEPreInit = {
-	// Without this, TinyMCE tries to determine its URL by looking at the
-	// <script> tag where it was loaded from, which of course fails here.
-	baseURL: 'about:blank',
-};
-
-global.window.setImmediate = function ( callback ) {
-	return setTimeout( callback, 0 );
-};
-
-global.window.requestAnimationFrame = function requestAnimationFrame(
-	callback
-) {
-	// eslint-disable-next-line no-restricted-syntax
-	const randomDelay = Math.round( ( Math.random() * 1_000 ) / 60 );
-
-	return setTimeout( () => callback( Date.now() ), randomDelay );
-};
-
-global.window.cancelAnimationFrame = function cancelAnimationFrame( handle ) {
-	return clearTimeout( handle );
-};
-
-// Ignoring `options` argument since we unconditionally schedule this ASAP.
-global.window.requestIdleCallback = function requestIdleCallback( callback ) {
-	const start = Date.now();
-
-	return setTimeout(
-		() =>
-			callback( {
-				didTimeout: false,
-				timeRemaining: () => Math.max( 0, 50 - ( Date.now() - start ) ),
-			} ),
-		0
-	);
-};
-
-global.window.cancelIdleCallback = function cancelIdleCallback( handle ) {
-	return clearTimeout( handle );
-};
-
-global.window.matchMedia = () => ( {
-	matches: false,
-	addListener: () => {},
-	addEventListener: () => {},
-	removeListener: () => {},
-	removeEventListener: () => {},
-} );
-
-// UserSettings global.
-global.window.userSettings = { uid: 1 };
+if ( typeof window !== 'undefined' ) {
+	// These are necessary to load TinyMCE successfully.
+	global.URL = window.URL;
+	global.window.tinyMCEPreInit = {
+		// Without this, TinyMCE tries to determine its URL by looking at the
+		// <script> tag where it was loaded from, which of course fails here.
+		baseURL: 'about:blank',
+	};
+
+	global.window.setImmediate = function ( callback ) {
+		return setTimeout( callback, 0 );
+	};
+
+	global.window.requestAnimationFrame = function requestAnimationFrame(
+		callback
+	) {
+		// eslint-disable-next-line no-restricted-syntax
+		const randomDelay = Math.round( ( Math.random() * 1_000 ) / 60 );
+
+		return setTimeout( () => callback( Date.now() ), randomDelay );
+	};
+
+	global.window.cancelAnimationFrame = function cancelAnimationFrame(
+		handle
+	) {
+		return clearTimeout( handle );
+	};
+
+	// Ignoring `options` argument since we unconditionally schedule this ASAP.
+	global.window.requestIdleCallback = function requestIdleCallback(
+		callback
+	) {
+		const start = Date.now();
+
+		return setTimeout(
+			() =>
+				callback( {
+					didTimeout: false,
+					timeRemaining: () =>
+						Math.max( 0, 50 - ( Date.now() - start ) ),
+				} ),
+			0
+		);
+	};
+
+	global.window.cancelIdleCallback = function cancelIdleCallback( handle ) {
+		return clearTimeout( handle );
+	};
+
+	global.window.matchMedia = () => ( {
+		matches: false,
+		addListener: () => {},
+		addEventListener: () => {},
+		removeListener: () => {},
+		removeEventListener: () => {},
+	} );
+
+	// UserSettings global.
+	global.window.userSettings = { uid: 1 };
+}
```

File: packages/stylelint-config/index.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 'use strict';
 
 module.exports = {
-	extends: 'stylelint-config-recommended',
+	extends: [ 'stylelint-config-recommended' ].map( require.resolve ),
```
Comment (gziolo) [inline packages/stylelint-config/index.js:5]: What's the reason for using `map` here` with a single item?

**Comment by gziolo** [comment]
I mostly skimmed the changes applied and it all looks like everything is finally on the right track. The most important part is that linting passes! Excellent work 👏🏻

**Comment by mikeybinns** [comment]
> https://jestjs.io/docs/configuration#testenvironment-string
> 
> Have you tried changing the test env for these test files in the package? That would be the correct environment anyway there. A code comment at the top of every failing test suite file should do the trick.
> 
> ```js
> /**
>  * @jest-environment node
>  */
> ```
Ah, yes I had tried this, but adding a eslint-disable rule about the tag being incorrect stopped it from working as it was no longer at the top of the file. This time, I tried adding an exclusion to the rule in the eslint config (just added to the monorepo root config) and it worked 👍

File: packages/stylelint-config/index.js (hunk 1 above)
Comment (mikeybinns) [reply packages/stylelint-config/index.js:5]: I made it an array when I was previously using `@stylistic/stylelint-config`, but when I removed that config, I kept it as an array.

There was then a second error where it wouldn't find the config, so I fixed that with map/require.resolve as that's how the other configs are set up.

I can convert this to a non-mapped version if you'd prefer that?

**Comment by mikeybinns** [comment]
@gziolo As expected, once I sorted the global jest things to be wrapped in window checks, I run into this error:
image

Stylelint is now an ESM package as of v16, which Jest doesn't like.

I tried converting only the test files to ESM, and then adding that flag to the test:unit command in the root package.json, but I now get this different error:
image

It's not always 'stylelint', sometimes it's other files e.g. `request for './hasLessInterpolation.js' is not yet fulfilled`

From what I can see, it's an issue with jest-runtime and we might be SOL unless you have any ideas.
The [Jest v30 alpha 1](https://github.com/jestjs/jest/releases/tag/v30.0.0-alpha.1) has this line:
```
[jest-runtime] Properly handle re-exported native modules in ESM via CJS
```
This indicates to me that maybe the new Jest version will fix this issue?

Note: I've pushed all my code so far again.

**Comment by mikeybinns** [comment]
@gziolo Actually, don't worry. I've found a way to use node exec to use the CLI to bypass using the node API entirely, and still get the same result. Probably is a bit clunky compared to using the API, but at least it works without having to rethink the whole jest setup 🥳

**Comment by mikeybinns** [comment]
All tests now pass after updating snapshots. I've checked the snapshot changes and all changes are correct/expected (e.g. just new properties available in v16 or changes to stylistic versions of rules) with the exception of 1 rule.

In the media queries test, there was a missing warning for `media-feature-name-no-unknown` for line 11 in `media-queries-invalid.css`. I'm not sure what change in the two stylelint versions means that this is different, but this rule no longer triggers with the `@media all and (max-width 699px) {` content. I believe this is down to it being a CSS syntax error as flagged by VS Code. I have verified the config has this rule enabled by changing this line to `@media all and (unknown) {` as per [the stylelint docs](https://stylelint.io/user-guide/rules/media-feature-name-no-unknown/), and with this change, the rule correctly flags a warning again for that file.

I believe that this is fine due to the css syntax error mentioned previously, however, if you feel it warrants further investigation, I can roll back that fix.

Other minor notes on the most recent commits:
- I've rolled back the ESM-ifying of the test files, they're back to CommonJS again.
- I corrected a misconfiguration in the scss-stylistic config so the scss rules override the base rules (something I wouldn't have found without these automated jest tests, so yay testing 🎉 )
- I've created a utils.cjs file to hold some common code, and specifically chosen the `.cjs` extension so Jest doesn't pick it up as a file with no tests. I couldn't see a clear way to make Jest ignore this file otherwise without changing the config, so this seemed the best option, however if there's precedent for this pattern in other packages, I'm happy to align it.

**Comment by gziolo** [comment]
I added missing labels to PR, and restarted one of the CI jobs with e2e tests. CI should become green after they finish 🎉  Excellent job taking it that far. I will give it a spin in the next days to see if my usual workflows work as expected.

> I've rolled back the ESM-ifying of the test files, they're back to CommonJS again.

Jest historically had troubles with adding proper support for ESM and I know it was very challenging for them due to how magic options they offered over years to mock the testing env to suite people needs 😀

> I couldn't see a clear way to make Jest ignore this file otherwise without changing the config, so this seemed the best option, however if there's precedent for this pattern in other packages, I'm happy to align it.

I had to double check the matching pattern but it’s fine to use subfolders:

https://github.com/WordPress/gutenberg/blob/47bf56ad0da4e733e98dcd53f205a973ebcda631/packages/jest-preset-default/jest-preset.js#L23

So `test/utils/index.js` should work fine. Such conventions are hard to discover 😅

I also plan to go over older PRs to let folks know and ask to help testing. Technically, it’s going to be breaking changes to two packages so we will have to include entries in the changelog files as explained in https://github.com/WordPress/gutenberg/tree/trunk/packages#maintaining-changelogs.

**Comment by gziolo** [comment]
When going through open PRs and issues, I noticed that this PR should also fix:
- https://github.com/WordPress/gutenberg/issues/54127

**Comment by mikeybinns** [comment]
Yep, fully expected to have to do changelog entries, I just wanted to ensure we had a finished PR before I did that 🙂

I'll update the utils convention and the changelog this afternoon but that shouldnt break anything.

**Comment by jacobcassidy** [comment]
@gziolo I would like to help test this by using it on a project. However, I'm unsure what steps are involved in replacing the `@wordpress/scripts` package downloaded via npm with this one. Can you give me an overview to point me in the right direction?

**Comment by mirka** [comment]
> I would like to help test this by using it on a project. However, I'm unsure what steps are involved in replacing the `@wordpress/scripts` package downloaded via npm with this one.

@jacobcassidy I think the easiest might be to take the package folder from the last commit in this branch, copy it into your project, and [modify your package.json](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#local-paths) so the dependency points to the path of the local copy?

**Comment by mikeybinns** [comment]
Copying the directory would work, but a cleaner way might be to use npm link.
Use Git to check out this PR's branch onto your machine.
Navigate in the command line to the `packages/scripts` folder (same location as the `@wordpress/scripts` package.json file)
Run `npm link` to create a link in node/npm
Go to your project and run `npm link --save-dev @wordpress/scripts`
This will install the scripts package you have locally as if it were a package in the project.
Note that your project package.json will now display a relative filepath for the scripts package: 
image

This is essentially a symlink, which means if you were to make a change to the scripts, it would be instantly applied to wherever it's linked (great for quick dev changes).

npm link docs: https://docs.npmjs.com/cli/v10/commands/npm-link

**Comment by gziolo** [comment]
My testing results from local copy.
- `npm install` locally doesn't apply changes ✅ 
- `npm run lint:css` passes ✅ 
- Stylistic modification get detected by `npm run lint:css`, example:
  
Screenshot 2024-09-12 at 10 47 06

I also executed Prettier on all SCSS files which isn't configured correctly for WP coding conventions, this is the result in this branch:
  
Screenshot 2024-09-12 at 10 53 06

The same exercise in the `trunk`:

Screenshot 2024-09-12 at 10 56 22
 
---

Aside, it looks like all these violations caused by Prettier run and reported are fixable. I'm wondering if it's more reliable than what ESLint offers for JavaScript. Anyway, that's a separate story whether to add support for (S)CSS to `wp-scripts format`.

File: packages/stylelint-config/index.js (hunk 1 above)
Comment (gziolo) [reply packages/stylelint-config/index.js:5]: I see it is used in other places with multiple items. It's fine.

**Review by gziolo — APPROVED** [review]
It looks good to me. Excellent work bringing it to the finish line!

I would love to see more testing before landing, but we can also publish the dev version of the package to npm and see how fast it works for fo.

**Comment by mikeybinns** [comment]
On the Prettier angle, that's exactly why I separated it out into separate configs. I personally use prettier on my personal projects, so I have no need for the formatting rules to be applied.

If you do go ahead and use prettier, you can switch the config in wp-scripts from scss-stylistic to scss and prevent the need to have something like [eslint-config-prettier](https://github.com/prettier/eslint-config-prettier) to disable all formatting rules. 🙂

Glad to see testing went well! 🥳

**Comment by gziolo** [comment]
I published all the updates included in this PR with the `@next` dist-tag to npm. You can test it in your projects with:

`npm install @wordpress/stylelint-config@next --save-dev`

See https://www.npmjs.com/package/@wordpress/stylelint-config/v/23.0.1-next.5368f64a9.0.

We have time until next Wednesday when the regular bi-weekly npm release happens to apply any necessary fixes if people report issues.

**Comment by jacobcassidy** [comment]
@gziolo The [`@next`](https://www.npmjs.com/package/@wordpress/stylelint-config/v/23.0.1-next.5368f64a9.0) package doesn't include the [`scss-stylistic.js`](https://github.com/mikeybinns/gutenberg/blob/754139f9a3c88aaed63af9f7d4f35dece9bdf39e/packages/stylelint-config/scss-stylistic.js) or [`stylistic.js`](https://github.com/mikeybinns/gutenberg/blob/754139f9a3c88aaed63af9f7d4f35dece9bdf39e/packages/stylelint-config/stylistic.js) files so when using the instructions to add:
```json
{
	"extends": [ "@wordpress/stylelint-config/stylistic" ]
}
```
or
```json
{
	"extends": [ "@wordpress/stylelint-config/scss-stylistic" ]
}
```
An error is thrown that those files don't exist.

**Comment by gziolo** [comment]
> @gziolo The [@next](https://www.npmjs.com/package/@wordpress/stylelint-config/v/23.0.1-next.5368f64a9.0) package doesn't include the [scss-stylistic.js](https://github.com/mikeybinns/gutenberg/blob/754139f9a3c88aaed63af9f7d4f35dece9bdf39e/packages/stylelint-config/scss-stylistic.js) or [stylistic.js](https://github.com/mikeybinns/gutenberg/blob/754139f9a3c88aaed63af9f7d4f35dece9bdf39e/packages/stylelint-config/stylistic.js) files so when using the instructions to add:

Thank you for testing. Great catch, I can confirm that when checking the content of the package:

https://unpkg.com/browse/@wordpress/stylelint-config@23.0.1-next.5368f64a9.0/

Screenshot 2024-09-13 at 11 29 17

Is anyone willing to contribute a change to the list of allowed files in the npm package?

https://github.com/WordPress/gutenberg/blob/632d8d62e8f8621c2771b266c1d2a68477bb3ca1/packages/stylelint-config/package.json#L26-L32

**Comment by mikeybinns** [comment]
> Is anyone willing to contribute...

I'm on it 👍

**Comment by mikeybinns** [comment]
Fix added here (didn't seem like I could add it to this PR anymore)
#65313

**Comment by gziolo** [comment]
Yes, new PR is the way to go. Reviewing 👍🏻

**Comment by gziolo** [comment]
New version is available on npm: https://www.npmjs.com/package/@wordpress/stylelint-config/v/23.0.1-next.1f6eadc42.0. I see both files when checking on [unpkg.com](https://unpkg.com/browse/@wordpress/stylelint-config@23.0.1-next.1f6eadc42.0/):

Screenshot 2024-09-13 at 12 55 16

**Comment by jeffpaul** [comment]
@vicobot-0815 mind sharing your WordPress.org username so I can ensure its included in the 6.7 credits listing?

**Pull request merged**